contracts/PeeranhaCommunityTokenFactory.sol

```diff
@@ -0,0 +1,174 @@
+      if (factoryData.factoryCommunitiesId[i] == communityId)
+        isAddedCommunityId = true;
+    }
+    if (!isAddedCommunityId) {
+      factoryData.factoryCommunitiesId.push(communityId);
+    }
+
+    uint256 contractsCommunityTokenLength = factoryData.peeranhaCommunitiesToken[communityId].length;
+    address communityTokenAddress = address(factoryData.peeranhaCommunitiesToken[communityId][contractsCommunityTokenLength - 1]);
+    emit CommunityTokenCreated(communityTokenAddress, communityId);
+  }
+
+  // This is to support Native meta transactions
+  // never use msg.sender directly, use _msgSender() instead
+  function _msgSender()
+      internal
+      override
+      view
+      returns (address sender)
+  {
+      return NativeMetaTransaction._msgSender();
+  }
+
+  function updateCommunityRewardSettings(address userAddress, uint32 communityId, address communityTokenContractAddress, uint256 maxRewardPerPeriod, uint256 activeUsersInPeriod) external override {
+    dispatcherCheck(userAddress);
+    IPeeranhaCommunityToken peeranhaCommunityToken = getContractCommunityToken(communityId, communityTokenContractAddress);
+    peeranhaCommunityToken.updateCommunityRewardSettings(maxRewardPerPeriod, activeUsersInPeriod);
+
+    emit CommunityRewardSettingsUpdated(communityTokenContractAddress);
+  }
+  
+  // set pools
+  function setTotalPeriodRewards(uint16 period) external override {
+    require(RewardLib.getPeriod() > period + 1, "period_not_ended");  // todo: tests
+    require(factoryData.peeranhaUser.isProtocolAdmin(_msgSender()), "not_allowed_not_protocal_admin");  // todo: tests
+
+    require(!factoryData.isSetPool[period], "pools_already_set");    // todo: tests
+    factoryData.isSetPool[period] = true;
+    uint256 rewardCommunitiesLength = factoryData.factoryCommunitiesId.length;
+    for (uint256 i; i < rewardCommunitiesLength; i++) {
+      RewardLib.PeriodRewardShares memory periodRewardShares = factoryData.peeranhaUser.getPeriodCommunityRewardShares(period, factoryData.factoryCommunitiesId[i]);
+      IPeeranhaCommunityToken[] memory contractsCommunityToken = getContractsCommunityToken(factoryData.factoryCommunitiesId[i]);
+      uint256 contractsCommunityTokenLength = contractsCommunityToken.length;
+      for (uint256 communityTokenIndex; communityTokenIndex < contractsCommunityTokenLength; communityTokenIndex++) {
+        IPeeranhaCommunityToken peeranhaCommunityToken = contractsCommunityToken[communityTokenIndex];
+        if (address(peeranhaCommunityToken) != address(0)) {
+          peeranhaCommunityToken.setTotalPeriodReward(periodRewardShares, period);
+        }
+      }
+    }
+    emit SetCommunityTokenPool(period);
+  }
+
+  function payCommunityRewards(address userAddress, uint16 period) external override {
+    dispatcherCheck(userAddress);
+    require(!factoryData.statusRewardContainer.statusReward[userAddress][period].isPaid, "reward_already_picked_up.");
```

Comment:
There is a big issue. User can call `claimRewards` to token rewards contract directly and this flag will not be checked.